helix-term/src/commands/lsp.rs

```diff
@@ -576,20 +579,57 @@ pub fn signature_help(cx: &mut Context) {
 
     cx.callback(
         future,
-        move |_editor, _compositor, response: Option<lsp::SignatureHelp>| {
-            if let Some(signature_help) = response {
-                log::info!("{:?}", signature_help);
-                // signatures
-                // active_signature
-                // active_parameter
-                // render as:
-
-                // signature
-                // ----------
-                // doc
-
-                // with active param highlighted
-            }
+        move |editor, compositor, response: Option<lsp::SignatureHelp>| {
+            let response = match response {
+                Some(s) => s,
+                None => {
+                    compositor.remove(SignatureHelp::ID);
+                    return;
+                }
+            };
+            let doc = doc!(editor);
+            let language = doc
+                .language()
+                .and_then(|scope| scope.strip_prefix("source."))
+                .unwrap_or("");
+
+            // If there are no signatures, response will be None, therefore
+            // at least one signature exists.
+            let signature = &response.signatures[response.active_signature.unwrap_or(0) as usize];
+            let mut contents = SignatureHelp::new(
+                signature.label.clone(),
+                language.to_string(),
+                Arc::clone(&editor.syn_loader),
+            );
+
+            let signature_doc = signature.documentation.as_ref().map(|doc| match doc {
+                lsp::Documentation::String(s) => s.clone(),
+                lsp::Documentation::MarkupContent(markup) => markup.value.clone(),
+            });
+            contents.set_signature_doc(signature_doc);
+
+            let active_param_range = || -> Option<(usize, usize)> {
+                let param_idx = signature
+                    .active_parameter
+                    .or(response.active_parameter)
+                    .unwrap_or(0) as usize;
+                let param = signature.parameters.as_ref()?.get(param_idx)?;
+                match &param.label {
+                    lsp::ParameterLabel::Simple(string) => {
+                        let start = signature.label.find(string.as_str())?;
+                        Some((start, start + string.len()))
+                    }
+                    lsp::ParameterLabel::LabelOffsets([start, end]) => {
+                        Some((*start as usize, *end as usize))
+                    }
+                }
+            };
+            contents.set_active_param_range(active_param_range());
+
+            let mut popup = Popup::new(SignatureHelp::ID, contents);
+            let old_popup = compositor.find_id::<Popup<SignatureHelp>>(SignatureHelp::ID);
+            popup.set_position(old_popup.and_then(|p| p.position().copied()));
+            compositor.replace_or_push(SignatureHelp::ID, popup);
```

Comment:
Interesting approach, I thought we would have to update signature_help in place and execute requests there, but this is cleaner 👍🏻

Reply:
How are you tracking `isRetrigger` though?

Reply:
Retriggers are not tracked at all right now, but all trigger characters are considered as retriggers in the spec. Trigger characters usually have `,` in them so it acts like a retrigger.